x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/role_mappings_table.tsx

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { Fragment, useState } from 'react';
+
+import {
+  EuiFieldSearch,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiIconTip,
+  EuiSpacer,
+  EuiTable,
+  EuiTableBody,
+  EuiTableHeader,
+  EuiTableHeaderCell,
+  EuiTableRow,
+  EuiTableRowCell,
+  EuiTextColor,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+import { ASRoleMapping } from '../../app_search/types';
+import { WSRoleMapping } from '../../workplace_search/types';
+import { EuiLinkTo } from '../react_router_helpers';
+import { RoleRules } from '../types';
+
+import {
+  ANY_AUTH_PROVIDER,
+  ANY_AUTH_PROVIDER_OPTION_LABEL,
+  ROLE_LABEL,
+  ALL_LABEL,
+  AUTH_PROVIDER_LABEL,
+  EXTERNAL_ATTRIBUTE_LABEL,
+  ATTRIBUTE_VALUE_LABEL,
+  FILTER_ROLE_MAPPINGS_PLACEHOLDER,
+  MANAGE_ROLE_MAPPING_BUTTON,
+} from './constants';
+
+interface IAccessItem {
+  name: string;
+}
+
+interface SharedRoleMapping extends ASRoleMapping, WSRoleMapping {
+  accessItems: IAccessItem[];
+}
+
+interface IRoleMappingsTableProps {
+  accessItemKey: 'groups' | 'engines';
+  accessHeader: string;
+  roleMappings: ASRoleMapping[] | WSRoleMapping[];
+  addMappingButton: React.ReactNode;
+  accessAllEngines?: boolean;
+  shouldShowAuthProvider?: boolean;
+  getRoleMappingPath(roleId: string): string;
+}
+
+const MAX_CELL_WIDTH = 24;
+
+const noItemsPlaceholder = <EuiTextColor color="subdued">&mdash;</EuiTextColor>;
+
+const getAuthProviderDisplayValue = (authProvider: string) =>
+  authProvider === ANY_AUTH_PROVIDER ? ANY_AUTH_PROVIDER_OPTION_LABEL : authProvider;
+
+export const RoleMappingsTable: React.FC<IRoleMappingsTableProps> = ({
+  accessItemKey,
+  accessHeader,
+  roleMappings,
+  addMappingButton,
+  getRoleMappingPath,
+  shouldShowAuthProvider,
+}) => {
+  const [filterValue, updateValue] = useState('');
+
+  // This is needed because App Search has `engines` and Workplace Search has `groups`.
+  const standardizeRoleMapping = (roleMappings as SharedRoleMapping[]).map((rm) => {
```

Comment:
The typing here is a little bit hairy with all of this casting. You may be able to simplify things if you change the type of `roleMappings` to `(ASRoleMapping | WSRoleMapping)[]`. 

It's *technically* a different type, that's a bit looser, but I find it's easier to work with and is still valid:

```diff
diff --git a/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/role_mappings_table.tsx b/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/role_mappings_table.tsx
index 5754b59c9e0..0acbd9f31bf 100644
--- a/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/role_mappings_table.tsx
+++ b/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/role_mappings_table.tsx
@@ -51,7 +51,7 @@ interface SharedRoleMapping extends ASRoleMapping, WSRoleMapping {
 interface IRoleMappingsTableProps {
   accessItemKey: 'groups' | 'engines';
   accessHeader: string;
-  roleMappings: ASRoleMapping[] | WSRoleMapping[];
+  roleMappings: Array<ASRoleMapping | WSRoleMapping>;
   addMappingButton: React.ReactNode;
   accessAllEngines?: boolean;
   shouldShowAuthProvider?: boolean;
@@ -76,8 +76,8 @@ export const RoleMappingsTable: React.FC<IRoleMappingsTableProps> = ({
   const [filterValue, updateValue] = useState('');
 
   // This is needed because App Search has `engines` and Workplace Search has `groups`.
-  const standardizeRoleMapping = (roleMappings as SharedRoleMapping[]).map((rm) => {
-    const _rm = { ...rm } as SharedRoleMapping;
+  const standardizeRoleMapping = roleMappings.map((rm) => {
+    const _rm = { ...rm };
     _rm.accessItems = rm[accessItemKey];
     return _rm;
   });

```